exchange/auction_test.go

```diff
@@ -94,92 +112,131 @@ func loadCacheSpec(filename string) (*cacheSpec, error) {
+				},
+				{
+					Min:       10,
+					Max:       20,
+					Increment: 0.5,
+				},
+			},
+		},
+		includeWinners:    specData.TargetDataIncludeWinners,
+		includeBidderKeys: specData.TargetDataIncludeBidderKeys,
+		includeCacheBids:  specData.TargetDataIncludeCacheBids,
+		includeCacheVast:  specData.TargetDataIncludeCacheVast,
+	}
+
 	testAuction := &auction{
-		winningBids:         winningBids,
+		winningBids:         winningBidsByImp,
 		winningBidsByBidder: winningBidsByBidder,
 		roundedPrices:       roundedPrices,
 	}
-	_ = testAuction.doCache(ctx, cache, bids, vast, &specData.BidRequest, 60, &specData.DefaultTTLs, bidCategory)
-	found := 0
-
-	for _, cExpected := range specData.ExpectedCacheables {
-		for _, cFound := range cache.items {
-			// make sure Data section matches exactly
-			if !vast {
-				// not testing VAST XML for custom cache key test
-				eq := jsonpatch.Equal(cExpected.Data, cFound.Data)
-				if !eq {
-					continue
-				}
-			}
+	_ = testAuction.doCache(ctx, cache, targData, &specData.BidRequest, 60, &specData.DefaultTTLs, bidCategory)
 
-			// make sure Key value is as expected
-			keymatch := false
-			if len(cExpected.Key) > 0 {
-				// can only verify prefix; remainder is random
-				if strings.HasPrefix(cFound.Key, cExpected.Key) {
-					keymatch = true
-				}
+	if len(specData.ExpectedCacheables) > len(cache.items) {
+		t.Errorf("%s:  [CACHE_ERROR] Less elements were cached than expected \n", fileDisplayName)
+	} else if len(specData.ExpectedCacheables) < len(cache.items) {
+		t.Errorf("%s:  [CACHE_ERROR] More elements were cached than expected \n", fileDisplayName)
+	} else { // len(specData.ExpectedCacheables) == len(cache.items)
+		// We cached the exact number of elements we expected, now we compare them side by side in n^2
+		var matched int = 0
+		var formattedExpectedData string
+		for i := 0; i < len(specData.ExpectedCacheables); i++ {
+			if specData.ExpectedCacheables[i].Type == prebid_cache_client.TypeJSON {
+				ExpectedData := strings.Replace(string(specData.ExpectedCacheables[i].Data), "\\", "", -1)
+				ExpectedData = strings.Replace(ExpectedData, " ", "", -1)
+				formattedExpectedData = ExpectedData[1 : len(ExpectedData)-1]
 			} else {
-				if len(cFound.Key) == 0 {
-					// Key is expected to be empty
-					keymatch = true
-				}
-			}
-			if !keymatch {
-				continue
+				formattedExpectedData = string(specData.ExpectedCacheables[i].Data)
 			}
-
-			// make sure TTLSeconds section matches exactly
-			if cExpected.TTLSeconds == cFound.TTLSeconds {
-				found++
+			for j := 0; j < len(cache.items); j++ {
+				if formattedExpectedData == string(cache.items[j].Data) &&
+					specData.ExpectedCacheables[i].TTLSeconds == cache.items[j].TTLSeconds &&
+					specData.ExpectedCacheables[i].Type == cache.items[j].Type {
+					if specData.ExpectedCacheables[i].Type == prebid_cache_client.TypeJSON {
+						matched++
+					}
+					if specData.ExpectedCacheables[i].Type == prebid_cache_client.TypeXML &&
```

Comment:
The above four lines should be omitted.  Regardless of the "type", we should check the key value.  Otherwise, we will miss cases where we incorrectly get a key for JSON or, if keys are ever possible for anything other than XML, we'll fail to check their values.  Simply remove the above four lines and insert an "if" at the start of the following line.

Reply:
Didn't look that far ahead to when we implement anything other than XML, we should also be aware that, in that hypothetical scenario, the else branch in line 207 could also need modifications. Anyways. I guess that discussion shall be left for that day, in the meantime, the correction has been made
```
201         for i := 0; i < len(specData.ExpectedCacheables); i++ {
202             if specData.ExpectedCacheables[i].Type == prebid_cache_client.TypeJSON {
203                 ExpectedData := strings.Replace(string(specData.ExpectedCacheables[i].Data), "\\", "", -1)
204                 ExpectedData = strings.Replace(ExpectedData, " ", "", -1)
205                 formattedExpectedData = ExpectedData[1 : len(ExpectedData)-1]
206             } else {
207                 formattedExpectedData = string(specData.ExpectedCacheables[i].Data)
208             }
```